AppController/djinn.rb

```diff
@@ -5081,6 +5071,67 @@ def find_lowest_free_port(starting_port, ending_port=0, appid="")
     return -1
   end
 
+  # This function checks if any node is to be considered down, and take
+  # appropriate actions. For example, if an appengine node is considered
+  # down, the AppServers running on it, must be considered lost.
+  def check_nodes_health()
+    if @all_stats.empty?
+      Djinn.log_debug("Cannot access node health since I don't have stats.")
+      return
+    end
+
+    @nodes.each { |node|
+      node_ip = node.private_ip
+      next if node_ip == my_node.private_ip
+      healthy = false
+
+      @all_stats.each{ |stats|
+        if stats['private_ip'] == node_ip
+          healthy = true
+          break
+        end
+      }
+      next if healthy
+
+      # Here we try to talk to the AC directly: if we get any response we
+      # considered the node up and running, otherwise we considered it
+      # crashed and we take actions.
+      Djinn.log_debug("Node at #{node_ip} is not in @all_stats!")
+      acc = AppControllerClient.new(node_ip, @@secret)
+      tries = RETRIES
+      begin
+        if acc.is_done_initializing? == true
+          Djinn.log_debug("The node at #{node_ip} is initialized.")
+          next
+        end
+      rescue FailedNodeException => except
+        tries -= 1
+        if tries > 0
+          Djinn.log_debug("node at #{node_ip} not responding: retrying.")
+          retry
+        end
+      end
+
+      Djinn.log_warn("Node at #{node_ip} is not responsive.")
+      to_terminate = {}
+      @app_info_map.each { |app, info|
+        next if not info['appengine']
```

Comment:
No: anytime we add an entry for an app, we add also some app information (ie language etc...).